ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -709,40 +909,100 @@ IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, long operationTimeMs)
    * @param operationTimeMs the time of the update operation
    * @param info this needs to be non-null in the case of recovery. Can be {@code null} otherwise. Used if the PUT
    *             record could not be found
+   * @param lifeVersion lifeVersion of this ttlUpdate record.
    * @return the {@link IndexValue} of the ttl update record
    * @throws StoreException if there is any problem writing the index record
    */
-  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs)
-      throws StoreException {
+  private IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs,
+      short lifeVersion) throws StoreException {
     validateFileSpan(fileSpan, true);
+    boolean hasLifeVersion = IndexValue.hasLifeVersion(lifeVersion);
     IndexValue value = findKey(id);
     if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+    }
+    short retrievedLifeVersion = value == null ? info.getLifeVersion() : value.getLifeVersion();
+    if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
       throw new StoreException("Id " + id + " deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
     } else if (value != null && value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
       throw new StoreException("TTL of " + id + " already updated in index" + dataDir, StoreErrorCodes.Already_Updated);
+    } else if (hasLifeVersion && retrievedLifeVersion > lifeVersion) {
+      throw new StoreException("LifeVersion conflict in index. Id " + id + " LifeVersion: " + retrievedLifeVersion
+          + " Undelete LifeVersion: " + lifeVersion, StoreErrorCodes.Life_Version_Conflict);
     }
     long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
     IndexValue newValue;
+
     if (value == null) {
       // It is possible that the PUT has been cleaned by compaction
+      // but the TTL update is going to still be placed?
```

Comment:
I have same question here, do you have any idea why TTL update is placed even when PUT has been cleaned up?

Reply:
i don't think it's possible. But I will take another look at later PR